Major Refactor for User VM, Linux Deamon and Nanvix Daemon

ppenna opened the pull request:
## Description

This PR refactors the User VM, Linux Daemon (linuxd) and Nanvix Daemon (nanvixd) to use tokio for driving synchronous and asynchronous communication.

Moving forward, this enables the single process co-exist with the multi-process architecture.

## AI Digest

This pull request migrates the build system and documentation from using the `microvm` binary to a new `uservm` binary. It updates all relevant build scripts, configuration files, and documentation to reflect this change, and cleans up related dependencies and build logic. Additionally, it improves argument validation and API ergonomics in the `linuxd` daemon and streamlines the CI test logic.

Key changes include:

**Migration from `microvm` to `uservm`:**

* The default workspace member and all references in `Cargo.toml` and `Makefile` are switched from `microvm` to `uservm`, including build, format, lint, and test targets. The `microvm` dependency is removed.
* Documentation and scripts are updated to reference `uservm.elf` instead of `microvm.elf`. 

**Build and test improvements:**

* The guest rlib test commands now explicitly enable the `std` feature for each package, improving test consistency. 
* The host cargo test command no longer forces the `std` feature by default, allowing more flexible feature selection. 

**CI and script enhancements:**

* The CI test logic in `scripts/ci.py` is improved to check for nanvixd test support before running tests, reducing redundant logic.

**Daemon and dependency updates:**

* The `mio` dependency is removed from `linuxd`, and `tokio` (with full features) is added, modernizing the async runtime.
* Argument parsing in `linuxd` is improved: error messages for missing mandatory arguments are more descriptive, and accessor methods are refactored to return references and provide sensible defaults.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -738,3 +370,116 @@ impl LinuxDaemon {
+    async fn get_user_vm_messages(
+        &self,
+        user_vm_connections: &mut HashMap<RawUserVmIdentifier, UserVmHandle>,
+    ) -> Result<Vec<(u32, UserVmHandle, Message)>, u32> {
```
The first `u32` should be refactored to a `RawUserVmIdentifier` in this method.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -738,3 +370,116 @@ impl LinuxDaemon {
+    /// Drain messages for each active user VM (best-effort, non-fatal on individual errors).
+    async fn process_user_vm_message(
+        &self,
+        uvm_id: u32,
```
Should be `RawUserVmIdentifier`.

ppenna commented on src/uservm/src/vmm/microvm/kvm/vcpu/mod.rs:
```diff
@@ -248,7 +186,7 @@ impl VirtualProcessor {
         vcpu_regs.rax = rax;
         vcpu_regs.rbx = rbx;
-        vcpu_regs.rflags = 2;
+        vcpu_regs.rflags = 2; // TODO: Have a constant for this.
```
Fixed.

ppenna commented on src/utils/nanvixd/src/cache/mod.rs:
```diff
@@ -171,6 +185,22 @@ impl SandboxCache {
                 );
                 self.sandbox_index.insert(tag.sandbox_id(), tag.clone());
+
+                // Wait for  sandbox.gatewauy_sockaddr() file to show up.
```
Fixed.

ppenna commented on src/utils/nanvix-bench/src/main.rs:
```diff
@@ -167,15 +181,15 @@ impl Benchmark {
     fn start_user_vm(&self, gateway_addr: Option<String>) -> Result<Child> {
         let mut user_vm_args: Vec<String> = vec![
             format!("{}/bin/microvm.elf", get_proj_root()),
```
Fixed.

csegarragonz commented on doc/run.md:
I don't know if this PR is the right place to address this, but in this document we should `s/MicroVM/UserVM/gc`

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -105,35 +96,55 @@ pub struct LinuxDaemon {
+        // an L2 VM or not.
+
+        let unbound_socket: UnboundSocket = UnboundSocket::new(self.control_plane_sockaddr_type);
+        match unbound_socket
```
A bit of a nitpick, but should we add a timeout to this `connect` and crash linuxd if we cannot connect to the control-plane instead of awaiting forever?

ppenna replied:
Fixed.

csegarragonz replied:
Sorry but where is it?

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -143,214 +154,55 @@ impl LinuxDaemon {
+        &self,
+        mut user_vm_stream: SocketStream,
+    ) -> Result<(UserVmIdentifier, UserVmHandle), Error> {
+        trace!("accepted connection from user VM (addr={user_vm_stream:?})",);
```
[nitpick] You are printing a SocketStream as if it were a socket address. Just making sure this is intentional. (I don't know exactly how a SocketStream would be debug-printed).

ppenna replied:
It is intentional and prints useful information.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -143,214 +154,55 @@ impl LinuxDaemon {
+        let new_msg: user_vm_api::NewUserVm = user_vm_api::NewUserVm::recv(&mut user_vm_stream)
+            .await
+            .map_err(|e| {
+                let reason = "failed to receive NewUserVm message from user VM";
```
`reason` is missing a type annotation.

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -398,281 +250,101 @@ impl LinuxDaemon {
-                                    .entry(uvm_id)
-                                    .or_default()
-                                    .push_back(worker_thread_handle);
+            tokio::select! {
```
You seem to assume that branches in the `select!` will be probed in top-down order, but I don't think that is true. If you want this behaviour you need to add:

```rust
tokio::select!{
    biased;
    ....
}
```

see the fairness section [1] on the tokio select docs.

[1] https://docs.rs/tokio/latest/tokio/macro.select.html#fairness

ppenna replied:
Yes, I am aware of that behavior. Number existed before my PR, I removed them.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -398,281 +250,101 @@ impl LinuxDaemon {
+                        },
+                    }
+                },
+                // 2. Accept at most one new user VM connection per loop.
```
Add a newline above the comment for consistency with the other blocks.

ppenna replied:
Fixed.

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -398,281 +250,101 @@ impl LinuxDaemon {
+                    let cmd: control_plane_api::NanvixdCommand = match result {
+                        Ok(cmd) => cmd,
+                        Err(e) => {
+                            error!("failed reading command from control-plane (error={e:?})");
```
[nitpick] Decompose in a re-usable `reason` that you can re-use.

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -398,281 +250,101 @@ impl LinuxDaemon {
+                    match cmd {
+                        control_plane_api::NanvixdCommand::Shutdown => {
+                            info!("linuxd received shutdown message from control-plane");
+                            let mut locked_worker_threads = worker_threads.lock().await;
```
[Nitpick] Type annotation.

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -398,281 +250,101 @@ impl LinuxDaemon {
+                result = self.user_vm_listener.accept() => {
+                    match result {
+                        Ok(user_vm_stream) => {
+                            let (user_vm_id, user_vm_handle) = self.accept_connections(user_vm_stream).await?;
```
[Nitpick] Type annotation.

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -398,281 +250,101 @@ impl LinuxDaemon {
+                            user_vm_connections.insert(user_vm_id, user_vm_handle);
+                        },
+                        Err(e) => {
+                            error!("error accepting connection from user VM (error={e:?})");
```
[Nitpick] Decompose into a reusable `let reason`

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -398,281 +250,101 @@ impl LinuxDaemon {
+                    let user_vm_messages = match user_vm_messages {
+                        Ok(msgs) => msgs,
+                        Err(uvm_id) => {
+                            user_vm_connections.remove(&uvm_id);
```
Should we add an error message here?

ppenna replied:
This is not an error. It happens if when the VM disconnects. I am adding a debug statement here.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -398,281 +250,101 @@ impl LinuxDaemon {
+                            message,
+                            worker_threads.clone(),
+                        ).await {
+                            error!("error processing message from user VM (uvm_id={uvm_id}, error={e:?})");
```
[Nitpick] Ditto about decomposing into a re-usable `let reason`.

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -398,281 +250,101 @@ impl LinuxDaemon {
         // by key.
-        let mut worker_threads: HashMap<RawUserVmIdentifier, VecDeque<WorkerThreadHandle>> =
-            HashMap::new();
+        let worker_threads: Arc<Mutex<HashMap<UserVmIdentifier, VecDeque<WorkerThreadHandle>>>> =
```
I am not able to see why we can get away without a `Mutex` in the `user_vm_connections` but we need one for the `worker_threads`?

ppenna replied:
`worker_threads` is used accross multiple `select!()` branches.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -398,281 +250,101 @@ impl LinuxDaemon {
                 }
+
+                // 3. Drain messages from existing user VM connections.
+                user_vm_messages = self.get_user_vm_messages(&mut user_vm_connections) => {
```
We cannot use `get_user_vm_messages` in a select branch because it is not cancellation safe. We need to re-think this design, I will expand more on the method itself below.

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -738,3 +407,116 @@ impl LinuxDaemon {
+            let Some(uvm_handle) = user_vm_connections.get(&uvm_id).cloned() else {
+                continue;
+            };
+            let message: Message = match Self::recv(uvm_handle.get_user_vm_reader()).await {
```
As mentioned above, I don't think this function is safe to be used in a `select!` statement. This is because it is not cancellation safe. See the tokio docs on cancellation safety [1].

Intuitively, any competing branches in the `select!` will be blocked on an `await`. Whenever one finishes, **all other futures will be dropped**. This means that any progress that has happened in that branch will be lost. This is fine when you are waiting on a socket or on a queue, but not on `get_user_vm_messages` as this is a stateful method.

For example, if another branch in the `select!` returns before we have iterated over all user vms (but after we have iterated over some), the messages that we have in the `messages` variable will be dropped. This will make us silently drop messages.

I am not sure what is the exact design for this. I was looking around and we could either have a fan-in pattern where we spawn one task per user vm stream, or use a future stream where we yield messages, but, in short, I don't think this method is safe as it is.

[1] https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/args.rs:
```diff
@@ -267,8 +274,8 @@ impl Args {
-    pub fn user_vm_bind_socket_type(&self) -> Option<String> {
-        self.user_vm_bind_sockaddr_type.clone()
+    pub fn user_vm_bind_socket_type(&self) -> &str {
+        self.user_vm_bind_sockaddr_type.as_deref().unwrap_or("unix")
```
Should we use a constant for this? Like `DEFAULT_USER_VM_SOCKADDR_TYPE`?

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/user_vm_handle.rs:
```diff
@@ -5,18 +5,21 @@
-use ::syscomm::{
-    BlockingSocketStream,
+use ::tokio::sync::Mutex;
+use syscomm::{
```
[Nitpick} Add `::` in the `use` statment.

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/user_vm_handle.rs:
```diff
@@ -5,18 +5,21 @@
+    UnboundSocket,
 };
-use ::syslog::error;
+use syslog::trace;
```
[Nitpick} Add `::` in the `use` statment.

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/user_vm_handle.rs:
```diff
@@ -25,38 +28,111 @@ use ::syslog::error;
+    /// Writer half used by worker threads to send responses without contending with the reader.
+    user_vm_writer: Arc<Mutex<SocketStreamWriter>>,
+    /// Address of the gateway socket to connect to.
+    gateway_sockaddr: String,
```
Should we stick with either `gateway_` or `gw_` for consistency?

ppenna replied:
Gateway. Fixed the code base accordingly.

csegarragonz commented on src/daemons/linuxd/src/worker_thread.rs:
```diff
@@ -256,28 +264,39 @@ impl WorkerThreadHandle {
-        let gw_stream: Option<Arc<Mutex<BlockingSocketStream>>> = uvm_handle.get_gw_vm_stream();
+        let uvm_stream: Arc<Mutex<SocketStreamWriter>> = uvm_handle.get_user_vm_writer();
+
+        let (gw_reader, gw_writer) = match Handle::current().block_on(uvm_handle.get_gw_vm_stream())
```
Am I correct in assuming that the gateway listener socket will only be initialized the first time we call handle_message on this user VM stream?

I don't know if this accounts for a situation where the client wants to immediately push something to the gateway socket? IIUC the client will only be able to conenct to the gateway socket the first time we call handle_message for a given user VM, whenever that may be.

csegarragonz replied:
I now see how this actually works because in nanvixd we block until we see that the gateway listener socket has been created, but I don't know if this approach is a good long-term solution. I think the read/write streams can be created lazily, but the listener socket should be created right when the user VM starts, as otherwise we cannot safely return from the `NEW` HTTP call.

csegarragonz commented on src/libs/control-plane-api/src/lib.rs:
```diff
@@ -27,170 +85,217 @@ use ::serde::{
+        async move {
+            // Read command sender.
+            let mut sender_bytes: [u8; mem::size_of::<Sender>()] = [0u8; mem::size_of::<Sender>()];
+            stream.read_exact(&mut sender_bytes).await?;
```
Should we have a timeout here?

ppenna replied:
No, timeouts should be included in the consumer of this library, if needed.

csegarragonz commented on src/libs/control-plane-api/src/lib.rs:
```diff
@@ -27,170 +85,217 @@ use ::serde::{
+    ///
+    /// # Returns
+    ///
+    /// THis function returns a future that, when resolved, yields either:
```
Typo: This

ppenna replied:
Fixed.

csegarragonz commented on src/libs/control-plane-api/src/lib.rs:
```diff
@@ -27,170 +85,217 @@ use ::serde::{
-    }
+            // Read command length.
+            let mut length_bytes: [u8; mem::size_of::<u32>()] = [0u8; mem::size_of::<u32>()];
+            stream.read_exact(&mut length_bytes).await?;
```
Just to confirm, the only reason why read_exact would fail is if we have an unrecoverable error rigth?

I am just trying to assert that this indeed closes #1048

ppenna replied:
Yes, it does.

csegarragonz commented on src/libs/syscomm/src/tests/tcp.rs:
```diff
@@ -0,0 +1,74 @@
+
+// Use tokio test runtime for async tests.
+#[tokio::test]
+async fn tcp_connect_bind_success() {
```
[Nitpick] This method is missing type annotations.

ppenna replied:
Fixed.

csegarragonz commented on src/libs/syscomm/src/tests/unix.rs:
```diff
@@ -0,0 +1,66 @@
+
+#[tokio::test]
+async fn unix_socket_success() {
+    // Create a unique path in temp dir based on pid and timestamp.
```
[Nitpick] This method is missing type annotations.

ppenna replied:
Fixed.

csegarragonz commented on src/libs/syscomm/src/socket_listener.rs:
```diff
@@ -0,0 +1,124 @@
+                addr: _,
+            } => {},
+            SocketListener::Unix { listener: _, path } => {
+                // Remove underlying socket file, because as dropping the listener just closes the
```
[Nitpick] I would remove the `as` for clearer wording.

ppenna replied:
Fixed.

csegarragonz commented on src/libs/user-vm-api/src/lib.rs:
```diff
@@ -91,88 +232,111 @@ impl NewUserVm {
-                io::Error::new(io::ErrorKind::InvalidData, reason)
-            })?;
+    pub async fn recv(stream: &mut SocketStream) -> Result<Self> {
+        ::syslog::trace!("recv():");
```
[Nitpick] You are already importing `::sysloc::trace`.

ppenna replied:
Fixed.

csegarragonz reviewed (commented):
I am done reviewing the linuxd side of things, and the libraries. Looks very good!

My only concern is regarding the cancellation safety of the method we use to drain user VM messages in linuxd.

csegarragonz commented on src/utils/nanvix-bench/src/main.rs:
```diff
@@ -403,11 +464,15 @@ impl Benchmark {
-            gateway_stream.write_all(&payload)?;
-            gateway_stream.read_exact(&mut response_payload)?;
-            latencies.push(start.elapsed().as_micros());
+            let user_vm_id = {
```
Any reason why we need this additional scope?

ppenna replied:
Keeping the gateway connection prevents resources from being dropped on server side, which is intended.

csegarragonz commented on src/utils/nanvixd/src/cache/mod.rs:
```diff
@@ -171,6 +185,23 @@ impl SandboxCache {
                 self.sandbox_index.insert(tag.sandbox_id(), tag.clone());
+
+                // Wait for gateway to show up.
+                // FIXME: Change to work with TCP as well.
```
Can we create a tracking issue for this? I am planning on implementing a conrol-plane message from linuxd to the nanvixd that should address this.

ppenna replied:
Fixed using throw-away connection. Please follow up with improved implementation that you suggested.

csegarragonz commented on src/utils/nanvixd/src/sandbox/microvm.rs:
```diff
@@ -28,8 +24,16 @@ use ::tokio::{
 };
+use control_plane_api::{
+    NanvixdCommand,
+    WireCommand,
```
I guess the linter would have flagged this, but I don't see why we need to include `WireCommand`?

ppenna replied:
Trait operations must be included in the scope.

ppenna replied:
It is required in order to use trait implementation.

csegarragonz commented on src/utils/nanvixd/src/config.rs:
```diff
@@ -172,15 +172,16 @@ pub fn user_vm_sockaddr_builder(tmp_str: &str, tenant_id: &str, l2: bool) -> Res
+    let sandbox_id: u32 = sandbox_id.into();
     let unix_socket_name: String =
-        format!("{tmp_str}/{tenant_id}:gw-{sandbox_id}{UNIX_SOCKET_SUFFIX}");
+        format!("{tmp_str}/{tenant_id}:gw-{sandbox_id:?}{UNIX_SOCKET_SUFFIX}");
```
Do we need the question mark `?` ? I am worried this would have unintended consequences when creating the socket name.

ppenna replied:
Fixed.

csegarragonz commented on src/utils/nanvixd/src/http.rs:
```diff
@@ -47,6 +47,7 @@ use ::syslog::{
     error,
 };
 use ::tokio::sync::Mutex;
+use syslog::trace;
```
```rust
use ::syslog::trace;
```

ppenna replied:
Fixed.

csegarragonz reviewed (commented):
This is a review of `./src/utils`.

csegarragonz commented on src/uservm/src/main.rs:
```diff
@@ -0,0 +1,184 @@
+        UnboundSocket::new(SocketType::from_str(args.system_vm_socket_type())?);
+    let system_vm_stream: SocketStream = match unbound_socket
+        .connect(args.system_vm_addr().to_string())
+        .await
```
Should we add a timeout here?

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/main.rs:
```diff
@@ -0,0 +1,184 @@
+        UnboundSocket::new(SocketType::from_str(args.control_plane_socket_type())?);
+    let control_plane_stream: SocketStream = match unbound_socket
+        .connect(args.control_plane_addr().to_string())
+        .await
```
Should we add a timeout here?

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/main.rs:
```diff
@@ -0,0 +1,184 @@
+    )?;
+
+    // Run virtual machine and check exit status code.
+    let vmm_handle = UserVm::spawn(UserVmArgs {
```
Missing type annotation.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/lib.rs:
```diff
@@ -0,0 +1,580 @@
+//==================================================================================================
+
+pub mod args;
+
```
Any reason for all these extra newlines here?

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/lib.rs:
```diff
@@ -0,0 +1,580 @@
+
+pub mod args;
+
+use ::libc::c_int;
```
Why is this `use` statement alone here and not with the others below?

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/lib.rs:
```diff
@@ -0,0 +1,580 @@
+pub mod vmm;
+
+/// POSIX signal used to interrupt a running virtual CPU thread.
+pub const INTERRUPT_SIGNAL: c_int = libc::SIGUSR1;
```
I thought we usually define constants after imports?

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/lib.rs:
```diff
@@ -0,0 +1,580 @@
+pub mod pal;
+
+/// Default bounded channel capacity used for cross-thread communication primitives.
+pub const CHANNEL_CAPACITY: usize = 1024;
```
This constant is repeated with linuxd. Could we abstract it away to `src/libs/config`?

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/args.rs:
```diff
@@ -45,21 +46,21 @@ pub struct Args {
     /// Log directory.
     log_directory: String,
+    /// Socket address type of the system VM socket.
+    system_vm_socket_type: String,
```
Has the system VM socket become a mandatory CLI arg? Does this mean we don't support spawning user VMs that are not bound to linuxd?

ppenna replied:
Yes, you are correct.

csegarragonz commented on src/uservm/src/args.rs:
```diff
@@ -45,21 +46,21 @@ pub struct Args {
-    control_plane_addr: Option<String>,
-    /// Control-plane socket type.
-    control_plane_socket_type: Option<SocketType>,
+    control_plane_addr: String,
```
Has the control plane address become a mandatory CLI argument now? It used to be optional.

ppenna replied:
Yes it is mandatory now.

csegarragonz commented on src/uservm/src/main.rs:
```diff
@@ -0,0 +1,184 @@
+        },
+    };
+
+    // Spawn I/O thread if we have at least one external stream to monitor.
```
The command seems to indicate some optionality, but both the system VM and control plane streams are now mandatory fields. Either change the comment, or revert those streams to optional again.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/io_thread.rs:
```diff
@@ -0,0 +1,218 @@
+    /// Spawns a new I/O thread.
+    ///
+    /// # Description
+    ///
```
Description missing, seems to be poorly formatted and appear in line 51 above.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/io_thread.rs:
```diff
@@ -0,0 +1,218 @@
+    /// # Description
+    ///
+    /// Runs the I/O thread according to the state in the snapshotting protocol state machine.
+    ///
```
Missing parameters here.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/io_thread.rs:
```diff
@@ -0,0 +1,218 @@
+
+        let (mut system_vm_rx, mut system_vm_tx): (SocketStreamReader, SocketStreamWriter) =
+            system_vm_stream.split();
+        // Split control-plane stream so that we can evolve independent reader/writer tasks.
```
Maybe add this comment too for the system vm stream?

ppenna replied:
Fixed.

csegarragonz commented:
@ppenna This PR also closes #1025

csegarragonz commented on src/uservm/src/io_thread.rs:
```diff
@@ -0,0 +1,218 @@
+                    }
+                }
+
+                result = NanvixdCommand::recv(&mut control_plane_rx) => {
```
Same concern as I expressed before in the linuxd loop: `NanvixdCommand::recv` is not cancellation safe, so we cannot use it (as is) in a branch of a `select!` statement.

The reason why it is not cacnellation safe is because we call `await` three times (for each part of the wire format) and the task may be cancelled any of those three, potentially losing progress.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/lib.rs:
```diff
@@ -0,0 +1,580 @@
+    ///
+    async fn run(args: UserVmArgs) -> Result<u16> {
+        trace!("spawn()");
+        let (memory_thread_data_tx, vcpu_thread_stdin_rx) =
```
No type annotations here.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/lib.rs:
```diff
@@ -0,0 +1,580 @@
+            load_snapshot_fn(microvm.clone(), filename.clone()),
+        );
+
+        let orchestrator_thread_handle = orchestrator_thread.spawn();
```
Missing type annotation.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/lib.rs:
```diff
@@ -0,0 +1,580 @@
+
+        let orchestrator_thread_handle = orchestrator_thread.spawn();
+
+        let exit_code = match vmm_thread.await {
```
Missing type annotation.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/lib.rs:
```diff
@@ -0,0 +1,580 @@
+// Standalone Functions
+//==================================================================================================
+
+/// Obtains a buffered writer for the virtual machine's standard error device.
```
Why do we have this line above the description?

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/lib.rs:
```diff
@@ -0,0 +1,580 @@
+/// # Returns
+///
+/// On success, the function returns a buffered writer for the virtual machine's standard error
+/// stream. An error is returned when the target file cannot be created.
```
Missing newline.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/lib.rs:
```diff
@@ -0,0 +1,580 @@
+    Box::new(input)
+}
+
+/// Builds an output callback that forwards messages emitted by the virtual machine.
```
This block comment is not properly structured. Missing description and newline at the end.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/lib.rs:
```diff
@@ -0,0 +1,580 @@
+pub fn output_fn(queue: Sender<Message>) -> Box<StdoutFn> {
+    // Output function used for emulating I/O port writes.
+    #[cfg(not(feature = "hyperlight"))]
+    let output = move |vm: &Arc<Mutex<VirtualMemory>>, data| -> Result<()> {
```
Missing type annotation (same for hyperlight).

ppenna replied:
It is not possible type annotate closures.

csegarragonz commented on src/uservm/src/lib.rs:
```diff
@@ -0,0 +1,580 @@
+
+    #[cfg(feature = "hyperlight")]
+    let output = move |data: Vec<u8>| -> Result<i32, hyperlight_host::HyperlightError> {
+        let bytes = data.as_slice();
```
Missing type annotation.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/lib.rs:
```diff
@@ -0,0 +1,580 @@
+
+    #[test]
+    fn get_stderr_writer_creates_target_file() -> AnyResult<()> {
+        let (path_str, path_buf) = unique_log_path("stderr")?;
```
This function is missing type annotations.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/lib.rs:
```diff
@@ -0,0 +1,580 @@
+        file_path.push(format!("nanvix-uservm-missing-{nanos}"));
+        file_path.push("stderr.log");
+
+        let result = get_stderr_writer(Some(file_path.to_string_lossy().into_owned()));
```
Type annotation.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/memory_thread.rs:
```diff
@@ -0,0 +1,315 @@
+    pub fn spawn(self) -> ::tokio::task::JoinHandle<()> {
+        trace!("spawn()");
+        ::tokio::spawn(async move {
+            let mut data_rx = self.data_rx;
```
Type annotation here.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/memory_thread.rs:
```diff
@@ -0,0 +1,315 @@
+    }
+
+    #[tokio::test]
+    async fn test_message_forward_and_credit_increment() {
```
All tests in this file are missing type annotations.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/orchestrator.rs:
````diff
@@ -0,0 +1,830 @@
+    /// let handle = orchestrator.spawn();
+    /// // ... perform other async work ...
+    /// let result = handle.await?; // Propagate any error from the run loop.
+    /// ```
````
Missing newline.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/orchestrator.rs:
```diff
@@ -0,0 +1,830 @@
+
+                // SAFETY: we call pthread_kill on a non-zero TID that we have received from
+                // the VCPU thread after boot, so this is safe.
+                let pthread_id: pthread_t = self.vcpu_tid as pthread_t;
```
This shutdown mechanism does not work for hyperlight. Any reason why you have removed the `cfg` flags that were here before?

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/vmm/hyperlight/mod.rs:
```diff
@@ -0,0 +1,360 @@
+
+impl Vmm {
+    pub fn new(args: MicroVmArgs) -> Result<Self> {
+        let guest = Guest;
```
This method has very little type annotations but I guess you copy and pasted it?

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/vmm/microvm/mod.rs:
```diff
@@ -0,0 +1,668 @@
+        let mut vm: VmFd = kvm.create_vm()?;
+
+        let irqchip: IrqChip = IrqChip::new(&mut kvm, &mut vm)?;
+
```
Any reason for all these newlines here?

ppenna replied:
Fixed.

csegarragonz reviewed (commented):
I have two concerns wrt the changes in `./src/uservm`.

First, you claim that this PR closes #1009, but what it does is it removes the optionality of the I/O thread and control-plane, making them mandatory arguments for the user VM. Is this intentional? I agree it cloes #1009, but not in the way I would have expected. (I.e. I would have thought that the fields would remain optional, and we would propagate this optionality to the channels).

Second, in the I/O thread we have the same issue than in linuxd when we try to use a custom method that cannot be gracefuly cacnelled as part of a select! branch.

LucasGdosR commented on src/uservm/src/io_thread.rs:
```diff
@@ -0,0 +1,217 @@
+    ///
+    /// # Parameters
+    ///
+    /// - `system_vm_stream`: Optional connection to the system VM.
```
This parameter is no longer optional.

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/io_thread.rs:
```diff
@@ -0,0 +1,217 @@
+    /// # Parameters
+    ///
+    /// - `system_vm_stream`: Optional connection to the system VM.
+    /// - `memory_thread_waker`: Waker token to notify the memory thread when we send in data_tx.
```
Tokens are no longer parameters.

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/io_thread.rs:
```diff
@@ -0,0 +1,217 @@
+    /// - `memory_thread_waker`: Waker token to notify the memory thread when we send in data_tx.
+    /// - `data_rx`: User VM receiver.
+    /// - `data_tx`: User VM sender.
+    /// - `orchestrator`: Waker token to notify the orchestrator when we send in control_tx.
```
Tokens are no longer parameters.

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/io_thread.rs:
```diff
@@ -0,0 +1,217 @@
+    /// - `orchestrator`: Waker token to notify the orchestrator when we send in control_tx.
+    /// - `control_tx`: Command sender.
+    /// - `control_rx`: Response receiver.
+    /// - `control_plane_stream`: Optional connection to the control-plane.
```
This parameter is no longer optional.

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/io_thread.rs:
```diff
@@ -0,0 +1,217 @@
+    ///
+    /// A handle to the I/O thread.
+    ///
+    #[allow(clippy::too_many_arguments)]
```
This `#[allow(...)]` is no longer needed.

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/io_thread.rs:
```diff
@@ -0,0 +1,217 @@
+    ///
+    /// # Description
+    ///
+    /// Runs the I/O thread according to the state in the snapshotting protocol state machine.
```
This is a stale description that I added. Could you change it? The I/O thread behaves the same regardless of the state machine.

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/io_thread.rs:
```diff
@@ -0,0 +1,217 @@
+                        Some(response) => {
+                            match response {
+                                IoControlResponse::FlushInput => {
+                                    debug!("input flush completed");
```
This was the previous implementation. I can reimplement it after this refactor, if you'd prefer.
```
    fn flush_linuxd_input(&mut self) -> Result<()> {
        while self.try_receive_from_system_vm()? != Break(()) {
            // Keep looping until receiving from the system VM would block, which breaks the loop.
        }
        Ok(())
    }
```

ppenna replied:
This is no longer required. Messages are no longer buffered.

LucasGdosR commented on src/uservm/src/io_thread.rs:
```diff
@@ -0,0 +1,217 @@
+                                    control_tx.send(IoControlCommand::LinuxDaemonFlushed).await?;
+                                },
+                                IoControlResponse::FlushOutput => {
+                                    debug!("output flush completed");
```
This was the previous implementation. I can reimplement it after this refactor, if you'd prefer.
```
    fn flush_microvm_output(&mut self) -> Result<()> {
        while self.try_receive_from_microvm()? != Break(()) {
            // Keep looping until `data_rx` is empty, which breaks the loop.
        }
        while !self.outgoing.is_empty() {
            self.try_send_to_system_vm()?;
        }
        Ok(())
    }
```

ppenna replied:
This is no longer required. Messages are no longer buffered.

LucasGdosR reviewed (commented):
Reviewed the I/O thread.

LucasGdosR commented on src/uservm/src/memory_thread.rs:
```diff
@@ -0,0 +1,312 @@
+    Receiver,
+    Sender,
+};
+use syslog::trace;
```
```
use ::syslog::trace;
```

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/memory_thread.rs:
```diff
@@ -0,0 +1,312 @@
+    ///
+    /// # Parameters
+    ///
+    /// - `data_rx`: Receives data messages from the I/O thread.
```
These descriptions are in the wrong place. `spawn` takes no parameters. These are parameters to `new`, and fields of the struct. Also, `_control_tx` should lose its leading `_`.

ppenna replied:
Fixed.

LucasGdosR reviewed (commented):
Reviewed the memory thread. Skipped tests.

LucasGdosR commented on src/uservm/src/orchestrator.rs:
```diff
@@ -0,0 +1,832 @@
+// Types
+//==================================================================================================
+
+pub type PauseFn = dyn FnMut() -> ::std::pin::Pin<Box<dyn ::std::future::Future<Output = Result<()>> + Send>>
```
I noticed this was `Fn` previously, not `FnMut`. I'm not sure whether it's correct.

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/orchestrator.rs:
```diff
@@ -0,0 +1,832 @@
+///
+/// # Description
+///
+/// Auxiliary enum to disambiguate when we are breaking out of a function because there are no more
```
This no longer disambiguates anything, since there's no longer an `Empty` variant. Should we keep the `enum`, or use a `()` unit? Should the documentation change?

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/orchestrator.rs:
```diff
@@ -0,0 +1,832 @@
+    /// Thread ID of the vCPU thread.
+    vcpu_tid: u64,
+    // Channel that receives commands from the I/O thread.
+    io_control_rx: Option<Receiver<IoControlCommand>>,
```
This channel is now optional. Should the rustdoc explain why?

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/orchestrator.rs:
```diff
@@ -0,0 +1,832 @@
+    // Channel that receives commands from the I/O thread.
+    io_control_rx: Option<Receiver<IoControlCommand>>,
+    // Channel that sends commands to the I/O thread.
+    io_control_tx: Option<Sender<IoControlResponse>>,
```
This channel is now optional. Should the rustdoc explain why?

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/orchestrator.rs:
```diff
@@ -0,0 +1,832 @@
+        vcpu_control_rx: Receiver<VcpuControlResponse>,
+        vcpu_control_tx: Sender<VcpuControlCommand>,
+        pause_microvm: Box<PauseFn>,
+        resume_microvm: Box<ResumeFn>,
```
We're missing the `create_snapshot` callback.

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/orchestrator.rs:
```diff
@@ -0,0 +1,832 @@
+                if self.state == State::Paused {
+                    if let Err(error) = self
+                        .vcpu_control_tx
+                        .send(VcpuControlCommand::CreateSnapshot(filepath))
```
The main thread used to create the snapshot through a callback function instead of passing a message to the vCPU thread to create it. Was this change intended?

ppenna replied:
Yes, it did.

LucasGdosR commented on src/uservm/src/orchestrator.rs:
```diff
@@ -0,0 +1,832 @@
+                    }
+
+                    match self.vcpu_control_rx.recv().await {
+                        Some(VcpuControlResponse::SnapshotCreated) => {
```
This only makes sense if the vCPU thread is the one to create the snapshot. If the main thread creates it, then it shouldn't receive this message, only send it through the I/O thread as part of the previous match-statement.

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/orchestrator.rs:
```diff
@@ -0,0 +1,832 @@
+                    error!("pause_protocol(): {reason}");
+                    anyhow::bail!(reason)
+                },
+                Err(_) => {
```
Should we capture the `_` variable?

ppenna replied:
Fixed.

LucasGdosR reviewed (commented):
Reviewed the orchestrator. Skipped tests. If you'd like me to go over the tests, just say so and I'll prioritize it soon.

LucasGdosR commented on src/uservm/src/vmm/microvm/mod.rs:
```diff
@@ -0,0 +1,668 @@
+    },
+    task,
+};
+use sys::error::ErrorCode;
```
```
use ::sys::error::ErrorCode;
```

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/vmm/microvm/mod.rs:
```diff
@@ -0,0 +1,668 @@
+///
+/// # Description
+///
+/// A structure that represents a MicroVM.
```
Shouldn't it be a VMM instead of a MicroVM? Or maybe the struct should be named `MicroVm` rather than `Vmm`? The rest of the documentation mentions `MicroVm` and MicroVM.

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/vmm/microvm/mod.rs:
```diff
@@ -0,0 +1,668 @@
+// SAFETY: Every component of InteriorMicroVmHandle is Send + Sync:
+// Emulator contains:
+// - `VirtualMemory`, which is Send + Sync;
+// - `input` and `output`, which are boxed function pointers accessed from a single thread;
```
This safety doc is stale. The contents of `InteriorMicroVmHandle` are now different from what's in this safety doc.

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/vmm/microvm/mod.rs:
```diff
@@ -0,0 +1,668 @@
+    ///
+    /// Creates a MicroVM.
+    ///
+    /// # Parameters
```
The parameter is now `MicroVmArgs`, which are here: https://github.com/nanvix/nanvix/blob/enhancement-refactor/src/uservm/src/vmm/mod.rs

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/vmm/microvm/mod.rs:
```diff
@@ -0,0 +1,668 @@
+            Some(VcpuControlCommand::Resume) => Ok(()),
+            Some(VcpuControlCommand::CreateSnapshot(filepath)) => {
+                self.handle_create_snapshot(filepath).await?;
+                Ok(())
```
This is part of the change from the main thread creating the snapshot to the vCPU thread creating the snapshot.

Note: this introduces a bug. It changes the semantics of `CreateSnapshot`. It automatically resumes the MicroVM, and it does so without writing to the kernel memory that a pause is no longer requested.

We need to change the implementation of `CreateSnapshot` to overwrite the kernel's memory, so it can resume running, or wait for an actual `Resume` command, by putting this into a loop.

ppenna replied:
I will delegate this to an issue and assign it to you. Unfortunately, PR impacted the snapshot protocol. The loop solution does not seem to be right, as it also does not actually loop. https://github.com/nanvix/nanvix/issues/1054

LucasGdosR commented on src/uservm/src/vmm/microvm/mod.rs:
```diff
@@ -0,0 +1,668 @@
+        self.vmem.clone()
+    }
+
+    pub fn guest(&self) -> Arc<Mutex<Guest>> {
```
Missing documentation.
```
    ///
    /// # Description
    ///
    /// Returns a reference to the guest of the virtual machine.
    ///
    /// # Returns
    ///
    /// A reference to the guest of the virtual machine.
    ///
```

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/vmm/microvm/mod.rs:
```diff
@@ -0,0 +1,668 @@
+        (vmem_filepath, kvm_filepath)
+    }
+
+    async fn handle_pause(&mut self) -> Result<()> {
```
Missing documentation.
```
    ///
    /// # Description
    ///
    /// Acknowledges a pause request and waits for the next command, either `Resume` or `CreateSnapshot`.
    ///
    /// # Returns
    ///
    /// Upon success, return empty. Otherwise, returns an error.
    ///
```

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/vmm/microvm/kvm/vcpu/mod.rs:
```diff
@@ -477,283 +424,156 @@ impl VirtualProcessor {
+    /// virtual processor state. Otherwise, it returns an error.
     ///
-    pub fn get_state(&self) -> Result<VirtualProcessorState> {
+    pub fn save_state(&self) -> Result<VirtualProcessorState> {
```
This is missing MSRs. I think you'd include it in `fpu.save_state()`, unless you understand that MSRs wasn't needed.

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/vmm/microvm/mod.rs:
```diff
@@ -0,0 +1,668 @@
+            },
+        }
+
+        let vcpu_state: VirtualProcessorState = self.vcpu.lock().await.save_state()?;
```
I went into this rabbithole and identified MSRs is no longer saved.

ppenna replied:
Fixed.

LucasGdosR commented on src/uservm/src/vmm/microvm/mod.rs:
```diff
@@ -0,0 +1,668 @@
+            },
+        };
+
+        if let Err(e) = self.vcpu.lock().await.load_state(&state) {
```
Likewise, I believe MSRs is probably not here either.

ppenna replied:
Fixed.

LucasGdosR reviewed (commented):
Reviewed what was previously the `MicroVm` and is now currently the `Vmm`.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -143,214 +156,84 @@ impl LinuxDaemon {
+            ::syscomm::SocketAddr::TCP_SOCKADDR_MAX_LEN
+        };
+        const NEW_USER_VM_MESSAGE_LEN: usize =
+            ::core::mem::size_of::<u32>() + ::core::mem::size_of::<u8>() + GATEWAY_SOCKADDR_MAX_LEN;
```
I feel this constant belongs to `user-vm-api`?

csegarragonz replied:
Also I don't think this is very portable in the sense that when we add a field to the NewUserVm struct we will need to update this constant. What is the issue with a lengh-prefixed wire-format? Is it handling partial reads?

ppenna replied:
I intentionally moved to fixed length message to simplify design.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -398,281 +281,133 @@ impl LinuxDaemon {
+                                match msg.cmd() {
+                                    control_plane_api::NanvixdCommand::Shutdown => {
+                                        info!("linuxd received shutdown message from control-plane");
+                                        let mut locked_worker_threads = worker_threads.lock().await;
```
Type annotation.

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -398,281 +281,133 @@ impl LinuxDaemon {
+                    let user_vm_messages = match user_vm_messages {
+                        Ok(msgs) => msgs,
+                        Err(uvm_id) => {
+                            debug!("run(): haversting connection to user VM (vm_id={uvm_id})");
```
Typo: `harvesting`. Interesting the spell-checker did not atch it...

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -738,3 +470,135 @@ impl LinuxDaemon {
+                        let err_msg: Message = crate::build_error(source, error.code);
+                        let uvm_writer: Arc<Mutex<SocketStreamWriter>> =
+                            uvm_handle.get_user_vm_writer();
+                        let mut writer_guard = uvm_writer.lock().await;
```
Type annotation.

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -738,3 +470,135 @@ impl LinuxDaemon {
+                        let err_bytes: [u8; config::kernel::IPC_MESSAGE_SIZE] = err_msg.to_bytes();
+                        let mut bytes_written: usize = 0;
+                        while bytes_written < err_bytes.len() {
+                            match writer_guard.write(&err_bytes[bytes_written..]).await {
```
Why no `write_all`? We are not in the `select!` arm (header) anymore, we are in the body.

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -143,214 +156,84 @@ impl LinuxDaemon {
+            ::core::mem::size_of::<u32>() + ::core::mem::size_of::<u8>() + GATEWAY_SOCKADDR_MAX_LEN;
+        let mut payload: [u8; NEW_USER_VM_MESSAGE_LEN] = [0u8; NEW_USER_VM_MESSAGE_LEN];
+        let mut payload_filled: usize = 0usize;
+        while payload_filled < payload.len() {
```
Why no `read_exact` here? We are not in the `select!` arm header anymore, we are in the body when we call this so we won't be randomly cancelled.

ppenna replied:
Fixed.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -738,3 +470,135 @@ impl LinuxDaemon {
+//==================================================================================================
+
+impl LinuxDaemon {
+    async fn get_user_vm_messages(
```
Did you consider an alternative design where we have one tokio task per connected user VM that continuously tries to recv from the stream and enqueues the message to a mpsc queue? The select! can then call recv() on the queue, and then drain it when it is awoken. Something like:

```rust
let (tx, mut rx) = tokio::sync::mpsc::channel(1024);
for (id, handle) in user_vm_connections.clone() {
    let tx = tx.clone();
    tokio::spawn(async move {
        loop {
            match Self::recv(handle.get_user_vm_reader()).await {
                Ok(m) => { let _ = tx.send((id, handle.clone(), m)).await; }
                Err(_) => break, // or report and continue
            }
        }
    });
}
```

It may be a pre-optimization, so we can also leave it for the future, but this would remove the need of reading messages one at a time, as in the select statment we can call `recv().await` once from the channel and then `try_recv` until WouldBlock.

(I know you asked if I had design suggestions, but I thought more about it after we talked)

ppenna replied:
No I have not. Let's consider that in a different PR.

csegarragonz reviewed (commented):
Left some comments on `linuxd/mod.rs`. No cancellation-related race conditions as far as I can see. 

I think we may have been overly-cautious in some areas (e.g. using `read/write` without `_exact` when it is safe to do so) but this should not block the PR.

PS: AFAICT `*_exact` is not safe to use in the select! branch, but it is safe to be used in the body.

csegarragonz commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -398,281 +281,133 @@ impl LinuxDaemon {
+
+                        },
+                        Ok(n) => {
+                            control_plane_buffer_filled += n;
```
Sorry, another one. It does not seem to me that we ever set `control_plane_buffer_filled` to zero?

ppenna replied:
Fixed.

csegarragonz commented on src/libs/control-plane-api/src/tests.rs:
```diff
@@ -0,0 +1,173 @@
+// Copyright(c) The Maintainers of Nanvix.
```
Seems to me that this tests are out of date with the latest `control-plane-api` that does not include any sending/receiving functionality?

ppenna replied:
Fixed.

csegarragonz commented on src/libs/control-plane-api/src/lib.rs:
```diff
@@ -2,195 +2,154 @@
-//! - bytes: [u8; length]
+//! Control Plane API
+//!
+//! This library provides a structured wire protocol for control-plane messages exchanged between
```
I don't think this is true anymore?

ppenna replied:
It is. we are just presenting a higher-level API.

csegarragonz commented on src/utils/nanvixd/src/cache/mod.rs:
```diff
@@ -170,6 +189,40 @@ impl SandboxCache {
+
+                // Attempt to connect to the gateway socket.
+                let now: Instant = Instant::now();
+                loop {
```
We can do this in a tokio-friendly idiomatic way with:

```rust
timeout(Duration::from_secs(CONNECT_TIMEOUT_SECS), unbound_gateway_socket.clone().connect())
```

then you can match for:

```rust
Ok(Ok(stream))
Ok(Err(e)) => // error connecting
Err(e) => // timed-out
```

PS: I may have gotten the order of the arguments the other way around.

ppenna replied:
Because if the gateway dos not exist, it will fail immediately.

csegarragonz commented on src/utils/nanvixd/src/sandbox/linuxd.rs:
```diff
@@ -72,12 +78,22 @@ impl LinuxDaemon {
+        )
+        .await
+        {
+            Ok(Ok(stream)) => stream,
```
I see you have used `timeout` here. I wonder why not in my previous comment?

ppenna replied:
There were near 100 suggestions on this PR. I missed that. sorry.

csegarragonz reviewed (commented):
A few more things. Nothing major, maybe only take a look at the `timeout` one.

csegarragonz commented on src/uservm/src/vmm/hyperlight/guest.rs:
```diff
@@ -0,0 +1,74 @@
+impl Guest {
+    // Adds a credit to the virtual machine's credit pool.
+    pub fn add_credit(&mut self, vmem: &mut VirtualMemory) -> Result<()> {
+        let vmem = &mut vmem.manager;
```
I think this was just code re-org, but there are no type annotations in this method. (Feel free to ignore)

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/vmm/microvm/kvm/vcpu/msr.rs:
```diff
@@ -0,0 +1,128 @@
+        trace!("Saving MSR state");
+
+        // Build `Msrs` out of entries.
+        let msr_index_list = match kvm.get_msr_feature_index_list() {
```
Ditto. Type annotation, feel free to ignore.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/main.rs:
```diff
@@ -0,0 +1,249 @@
+// Standalone Functions
+//==================================================================================================
+
+async fn socket_write_all(stream: &mut SocketStream, buf: &[u8]) -> ::std::io::Result<()> {
```
Why can we not call directly `write_all` here? Maybe related to a previous comment I made before regarding cancellation-safety and when is it needed.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/io_thread.rs:
```diff
@@ -0,0 +1,265 @@
+                            let bytes: [u8; ::std::mem::size_of::<Message>()] = msg.to_bytes();
+                            let mut offset: usize = 0;
+                            while offset < bytes.len() {
+                                let written: usize = system_vm_tx.write(&bytes[offset..]).await.map_err(|e| {
```
As said before, I am pretty sure we can call `write_all` here.

ppenna replied:
Fixed.

csegarragonz commented on src/uservm/src/io_thread.rs:
```diff
@@ -0,0 +1,265 @@
+                            trace!("message forwarded to system VM")
+                        },
+                        None => {
+                            ::syslog::debug!("User VM channel closed, exiting I/O thread");
```
[Nitpick] You are already importing `debug`.

ppenna replied:
Fixed.

csegarragonz reviewed (commented):
LGTM! My only comments are related to the ones in linuxd regarding unnecessary cancel-safety. Happy to discuss those offline.

ppenna committed:
[daemons] E: Tokio in Sync and Async Comm

This commit refactors the User VM, Linux Daemon (linuxd) and Nanvix
Daemon (nanvixd) to use tokio for driving synchronous and asynchronous
communication.
